internal/lm/output.go

```diff
@@ -0,0 +1,192 @@
+)
+
+// Outputer defines a mechanism to output labels.
+type Outputer interface {
+	Output(Labels) error
+}
+
+// TODO: Replace this with functional options.
+func NewOutputer(config *spec.Config, nodeConfig flags.NodeConfig, clientSets flags.ClientSets) Outputer {
+	if config.Flags.UseNodeFeatureAPI == nil || !*config.Flags.UseNodeFeatureAPI {
+		return ToFile(*config.Flags.GFD.OutputFile)
+	}
+
+	return &nodeFeatureObject{
+		nodeConfig:   nodeConfig,
+		nfdClientset: clientSets.NFD,
+	}
+}
+
+func ToFile(path string) Outputer {
+	if path == "" {
+		return &toWriter{os.Stdout}
+	}
+
+	o := toFile(path)
+	return &o
+}
+
+// toFile writes to the specified file.
+type toFile string
+
+// toWriter writes to the specified writer
+type toWriter struct {
+	io.Writer
+}
+
+func (path *toFile) Output(labels Labels) error {
+	klog.Infof("Writing labels to output file %v", *path)
+
+	buffer := new(bytes.Buffer)
+	output := &toWriter{buffer}
+	if err := output.Output(labels); err != nil {
+		return fmt.Errorf("error writing labels to buffer: %v", err)
+	}
+	err := writeFileAtomically(string(*path), buffer.Bytes(), 0644)
+	if err != nil {
+		return fmt.Errorf("error atomically writing file '%s': %w", *path, err)
+	}
+	return nil
+}
+
+func (output *toWriter) Output(labels Labels) error {
+	for k, v := range labels {
+		_, err := fmt.Fprintf(output, "%s=%s\n", k, v)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func writeFileAtomically(path string, contents []byte, perm os.FileMode) error {
+	absPath, err := filepath.Abs(path)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve absolute path of output file: %v", err)
+	}
+
+	absDir := filepath.Dir(absPath)
+	tmpDir := filepath.Join(absDir, "gfd-tmp")
+
+	err = os.MkdirAll(tmpDir, os.ModePerm)
+	if err != nil && !os.IsExist(err) {
+		return fmt.Errorf("failed to create temporary directory: %v", err)
+	}
+	defer func() {
+		if err != nil {
+			os.RemoveAll(tmpDir)
+		}
+	}()
+
+	tmpFile, err := os.CreateTemp(tmpDir, "gfd-")
+	if err != nil {
+		return fmt.Errorf("fail to create temporary output file: %v", err)
+	}
+	defer func() {
+		if err != nil {
+			tmpFile.Close()
+			os.Remove(tmpFile.Name())
+		}
+	}()
+
+	err = os.WriteFile(tmpFile.Name(), contents, perm)
+	if err != nil {
+		return fmt.Errorf("error writing temporary file '%v': %v", tmpFile.Name(), err)
+	}
+
+	err = os.Rename(tmpFile.Name(), path)
+	if err != nil {
+		return fmt.Errorf("error moving temporary file to '%v': %v", path, err)
+	}
+
+	err = os.Chmod(path, perm)
+	if err != nil {
+		return fmt.Errorf("error setting permissions on '%v': %v", path, err)
+	}
+
+	return nil
+}
+
+const nodeFeatureVendorPrefix = "nvidia-features-for"
+
+type nodeFeatureObject struct {
+	nodeConfig   flags.NodeConfig
+	nfdClientset nfdclientset.Interface
+}
+
+// UpdateNodeFeatureObject creates/updates the node-specific NodeFeature custom resource.
+func (n *nodeFeatureObject) Output(labels Labels) error {
```

Comment:
I'll leave the func name `UpdateNodeFeatureObject` as it was before.

Reply:
This needs to be an `Outputer`. The point is that the caller doesn't care whether labels are output to a file or the node feature API.